### DISCUSSION TITLE
Slow and Inconsistent Site Replication in MinIO Operator Across Kubernetes Clusters

### DISCUSSION THREAD

**Discussion by nutsn507** [minio/minio#20775] [Q&A]
### Question: Slow and Inconsistent Site Replication in MinIO Operator Across Kubernetes Clusters

I’ve set up MinIO site replication using the MinIO Operator across two Kubernetes clusters. The source has 525 GiB of data spread across 1,401,821 objects. Below is the detailed object histogram and `mc stat` output for the source bucket:

```bash
bash-5.1$ mc stat minio-testrep/testrep
Name      : testrep
Date      : 2024-12-18 03:00:48 UTC
Size      : N/A
Type      : folder

Properties:
  Versioning: Enabled
  Replication: Enabled
  Location: us-east-1
  Anonymous: Disabled
  ILM: Disabled

Usage:
      Total size: 525 GiB
   Objects count: 1,401,821
  Versions count: 1,410,821

Object sizes histogram:
   1399456 object(s) BETWEEN_1024B_AND_1_MB
    958757 object(s) BETWEEN_1024_B_AND_64_KB
        65 object(s) BETWEEN_10_MB_AND_64_MB
        47 object(s) BETWEEN_128_MB_AND_512_MB
      1345 object(s) BETWEEN_1_MB_AND_10_MB
     44572 object(s) BETWEEN_256_KB_AND_512_KB
     11035 object(s) BETWEEN_512_KB_AND_1_MB
    385092 object(s) BETWEEN_64_KB_AND_256_KB
        21 object(s) BETWEEN_64_MB_AND_128_MB
       106 object(s) GREATER_THAN_512_MB
       781 object(s) LESS_THAN_1024_B
```

The replication is over an L2 network with 10 Gbps bandwidth. However, I’m facing two major issues:

1. **Replication is extremely slow.**
2. **The number of objects and their sizes on the target cluster are inconsistent.**

### Observations

1. `mc admin replicate status` shows the replication process received 475,809 objects totaling 975 GiB. However, when checking the target bucket, only 168,525 objects are present, with a total size of just 35 GiB. Below are the relevant command outputs:

   **Replication status:**
   ```bash
   bash-5.1$ mc admin replicate status minio-testrep
   Bucket replication status:
   ●  1/1 Buckets in sync

   Policy replication status:
   ●  5/5 Policies in sync

   User replication status:
   ●  3/3 Users in sync

   Group replication status:
   No Groups present

   ILM Expiry Rules replication status:
   No ILM Expiry Rules present

   Object replication status:
   Replication status since 19 hours
   10.4.37.40
   Replicated:    6,552 objects (0 B)
   Received:      475,809 objects (975 GiB)
   Queued:        ● 0 objects, (0 B) (avg: 1 objects, 0 B; max: 49 objects, 0 B)
   Workers:       0 (avg: 2; max 67)
   Transfer Rate: 0 B/s (avg: 0 B/s; max 0 B/s)
   Latency:       2ms (avg: 2ms; max 254ms)
   Link:          ● online (total downtime: 0 milliseconds)
   Errors:        0 in last 1 minute; 0 in last 1hr; 0 since uptime
   ```

   **Target bucket stats:**
   ```bash
   bash-5.1$ mc stat minio-testrep/testrep
   Name      : testrep
   Date      : 2024-12-18 03:30:32 UTC
   Size      : N/A
   Type      : folder

   Properties:
     Versioning: Enabled
     Replication: Enabled
     Location: us-east-1
     Anonymous: Disabled
     ILM: Disabled

   Usage:
         Total size: 35 GiB
      Objects count: 168,525
     Versions count: 172,354

   Object sizes histogram:
      163606 object(s) BETWEEN_1024B_AND_1_MB
      127072 object(s) BETWEEN_1024_B_AND_64_KB
           9 object(s) BETWEEN_10_MB_AND_64_MB
          14 object(s) BETWEEN_128_MB_AND_512_MB
         169 object(s) BETWEEN_1_MB_AND_10_MB
        5001 object(s) BETWEEN_256_KB_AND_512_KB
        1429 object(s) BETWEEN_512_KB_AND_1_MB
       30104 object(s) BETWEEN_64_KB_AND_256_KB
           2 object(s) BETWEEN_64_MB_AND_128_MB
          12 object(s) GREATER_THAN_512_MB
        4713 object(s) LESS_THAN_1024_B
   ```

2. While the replication status indicates a substantial amount of data has been received, the bucket stats do not reflect this progress.

3. No replication errors are reported in the logs, and the network link seems stable.

### Question

1. Why is there such a large discrepancy between the number of objects received and the objects present in the target bucket?
2. What could be causing the replication to proceed so slowly, despite the 10 Gbps network bandwidth?
3. Are there specific configurations or optimizations in the MinIO Operator, Kubernetes, or replication policies that could resolve this?

Any advice or insights into troubleshooting this would be greatly appreciated!

**Comment by harshavardhana** [comment]
Run `mc support perf site-replication alias/` to check your bandwidth between sites, to verify if your 10gbps requirement is correct

**Reply by nutsn507** [reply]
Thank you for the suggestion. Unfortunately, I am unable to run the mc support perf site-replication command because I don't have a subscription for the MinIO subnet. When I attempt to execute the command, I receive the following error:

```bash
bash-5.1$ mc support perf site-replication minio-testrep
mc: <ERROR>  Please register the cluster first by running 'mc license register minio-testrep'
```

If there are alternative methods to test bandwidth between the sites without requiring a subscription, I would greatly appreciate your guidance. Thank you!

**Comment by titic443** [comment]
I had the same issue.
